### PR TITLE
feat: add ERC20 swap operator

### PR DESCRIPTION
**Pull request by 0xble**
## Motivation

Parties need an elegant solution to propose swapping tokens held by their group. This has been difficult to do through previous proposals because often, by the time the proposal was executed, the proposal data (e.g. price) would become outdated.

## Solution

Build a new operator allowing for ERC20 token swaps which enables execution-time calldata containing time-sensitive data like the price. Update the existing `OperatorProposal` to be able to transfer more assets than just ETH to the operator to use on behalf of the Party. There is added support for ERC20s, ERC721s, and ERC1155s.

### CODE REVIEW

File: contracts/operators/ERC20SwapOperator.sol [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+
+    /// @notice Set the allowed targets that can be used to perform swaps. Can
+    ///         only be called by the PartyDAO multisig.
+    /// @param target The target contract address.
+    /// @param isAllowed Whether the target is allowed.
+    function setTargetAllowed(address target, bool isAllowed) external onlyPartyDao {
+        isTargetAllowed[target] = isAllowed;
+    }
```
Comment (arr00) [inline contracts/operators/ERC20SwapOperator.sol:84]: Thoughts on emitting an event here?

File: contracts/operators/ERC20SwapOperator.sol [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+contract ERC20SwapOperator is IOperator {
+    using LibAddress for address payable;
+    using LibRawResult for bytes;
+    using LibERC20Compat for IERC20;
+
+    event ERC20SwapOperationExecuted(
+        Party party,
+        IERC20 fromToken,
+        IERC20 toToken,
+        uint256 amount,
```
Comment (arr00) [inline contracts/operators/ERC20SwapOperator.sol:22]: why not add context here

File: contracts/operators/ERC20SwapOperator.sol [hunk 3]
```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.20;
+
+import "./IOperator.sol";
+import "../party/Party.sol";
+import "../tokens/IERC20.sol";
+import "../utils/LibAddress.sol";
+import "../utils/LibERC20Compat.sol";
+import "../utils/LibRawResult.sol";
```
Comment (arr00) [inline contracts/operators/ERC20SwapOperator.sol:9]: named imports?

File: contracts/proposals/OperatorProposal.sol [hunk 4]
```diff
@@ -36,13 +59,39 @@ contract OperatorProposal {
...
+            AssetData memory asset = data.assets[i];
+            if (asset.tokenType == OperatorTokenType.ETH) {
+                ethToTransfer += asset.amount;
+            } else if (asset.tokenType == OperatorTokenType.ERC20) {
+                IERC20(asset.token).compatTransfer(address(data.operator), asset.amount);
+            } else if (asset.tokenType == OperatorTokenType.ERC721) {
+                IERC721(asset.token).safeTransferFrom(
+                    address(this),
+                    address(data.operator),
+                    asset.tokenId
```
Comment (arr00) [inline contracts/proposals/OperatorProposal.sol:74]: assuming the operator is set by us, is a safe transfer needed?

**Review by arr00 — COMMENTED** [review]
Some comments so far

File: contracts/operators/ERC20SwapOperator.sol (hunk 2 above)
Comment (arr00) [reply contracts/operators/ERC20SwapOperator.sol:24]: This wasn't addressed

Comment (0xble) [reply contracts/operators/ERC20SwapOperator.sol:28]: Tis now

**Pull request merged**